### PR TITLE
HELM-4: Move event parameters into their own table

### PR DESCRIPTION
**Pull request by christianpape**
* JIRA: https://issues.opennms.org/browse/HELM-4

### CODE REVIEW

File: core/schema/src/main/liquibase/21.0.0/changelog.xml [hunk 1]
```diff
@@ -57,4 +57,54 @@
...
+        <constraints nullable="false"/>
+      </column>
+      <column name="type" type="VARCHAR(256)">
+        <constraints nullable="false"/>
+      </column>
+    </createTable>
+    <addPrimaryKey tableName="event_parameters"
+                   columnNames="eventid,name"
+                   constraintName="pk_eventparameters" />
+    <addForeignKeyConstraint baseTableName="event_parameters" baseColumnNames="eventid"
```
Comment (j-white) [inline core/schema/src/main/liquibase/21.0.0/changelog.xml]: Should we add another constraint to ensure that `(eventid, name)` is unique?

File: core/schema/src/main/liquibase/21.0.0/changelog.xml [hunk 2]
```diff
@@ -57,4 +57,54 @@
...
+                             constraintName="fk_eventparameterseventid"
+                             referencedTableName="events"
+                             referencedColumnNames="eventid"
+                             onDelete="CASCADE"/>
+
+    <dropColumn tableName="alarms" columnName="eventparms" />
+    <addColumn tableName="alarms">
```
Comment (j-white) [inline core/schema/src/main/liquibase/21.0.0/changelog.xml]: After discussing this with @pbrane, we think we should just always use `lasteventid` as the source of the parameters rather than introducing another column and constraint.

File: core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+    @Override
+    public boolean requiresOnmsRunning() {
+        return false;
+    }
+
+    @Override
+    public void preExecute() throws OnmsUpgradeException {
+        ;
+        // check for eventparms column
```
Comment (j-white) [inline core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java]: Is this still necessary given that we run the `preMigrationStatement` in the `execute` call?

File: core/schema/src/main/liquibase/21.0.0/changelog.xml (hunk 1 above)
Comment (fooker) [inline core/schema/src/main/liquibase/21.0.0/changelog.xml]: It's a primary key, which is, by definition, always unique.

File: core/schema/src/main/liquibase/21.0.0/changelog.xml (hunk 2 above)
Comment (fooker) [inline core/schema/src/main/liquibase/21.0.0/changelog.xml]: Can you summarize the arguments of the discussion? I've had the feeling that the `updateOnReduction` mechanism is somewhat strange for parameters anyway.

Should we add some kind of warning if there are configs which specify update on reduction for `parms`?

**Comment by pbrane** [comment]
> On Aug 23, 2017, at 05:05, Dustin Frisch <notifications@github.com> wrote:
> 
> It's a primary key, which is, by definition, always unique.

I think he was referring to the uniqueness of the ID and the Name fields combined as a composite key.

**Comment by pbrane** [comment]
> On Aug 23, 2017, at 05:08, Dustin Frisch <notifications@github.com> wrote:
> 
> Can you summarize the arguments of the discussion? I've had the feeling that the updateOnReduction mechanism is somewhat strange for parameters anyway.

The discussion centered around keeping the preservation of the same behavior in that a user could configure event reduction into an alarm whereby the parameters from the most recent event are placed into the content of the alarm vs keeping the parameters from the original alarm instantiating event.

We also discussed if we should continue allowing this configuration and just always updating the parameters with the most recent set from the last event.

I really don't recall the decision but the decision is really important when you consider that the auto-clean feature's intention is to only keep the most recent event in the DB after each reduction.

File: core/schema/src/main/liquibase/21.0.0/changelog.xml (hunk 1 above)
Comment (j-white) [inline core/schema/src/main/liquibase/21.0.0/changelog.xml]: Of course.

File: core/schema/src/main/liquibase/21.0.0/changelog.xml (hunk 2 above)
Comment (j-white) [inline core/schema/src/main/liquibase/21.0.0/changelog.xml]: This way we can keep this existing `auto-clean` logic, and avoid making additional references to the events.

This is a behavioral change though, and will effectively make `updateOnReduction` for `params` always `true`.

I don't think we need any kind of warning other than a mention in the release notes.

**Comment by j-white** [comment]
Looks good overall. I'll spend some more time testing the upgrade logic.

File: core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    }
+
+    @Override
+    public void execute() throws OnmsUpgradeException {
+        long eventCount = 0, parameterCount = 0;
+
+        try (final Connection connection = DataSourceFactory.getInstance().getConnection()) {
+
+            final Statement preMigrationStatement = connection.createStatement();
+            final ResultSet preMigrationResultSet = preMigrationStatement.executeQuery("SELECT EXISTS (SELECT 1 FROM information_schema.columns WHERE table_schema='public' AND table_name='events' AND column_name='eventparms')");
```
Comment (j-white) [inline core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java]: Should we take into account any cases where the `table_schema` is not `public`?

One way would be to omit the `table_schema` from the where clause, and search all schemas in the catalog.

**Comment by j-white** [comment]
I encountered the following while running the installer against a large database:
```
Processing EventParameterMigratorOffline: Moves event parameters from eventparms column to table
- Running pre-execution phase          
- Running execution phase                                                                                                                                     
  Warning: can't perform the upgrade operation because: Can't move event parameters to table: Java heap space
- Executing rollback phase             
- Running post-execution phase         
                                                                                                                                                              
Finished in 17 seconds
```

I confirmed that the installer is started using the options specified in `etc/opennms.conf`, in my case, I have:
```
ADDITIONAL_MANAGER_OPTIONS="-XX:+UseG1GC -Xmx2g"
```

We may need to the upgrade task to flush more frequently than every 10k records.

**Comment by j-white** [comment]
While working on some other code, I came across this line: https://github.com/OpenNMS/opennms/blob/b287fa31c6c64f1850a1564eb79141d107975fcc/opennms-services/src/main/java/org/opennms/netmgt/vacuumd/AutomationProcessor.java#L598

Here it appears that the `eventParms` are used in the action-event, if the field is returned by the trigger.

This originates from https://github.com/OpenNMS/opennms/commit/a20faedef3cd3b6f9b8ff0316abc930fd40e7d17, which is where the eventParms were introduced into the alarms table in the first place.

This functionality may no longer be required, none of the default automations use it, but we should make note of it somewhere.

File: core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java (hunk 4 above)
Comment (christianpape) [inline core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java]: At this point I can also use the ResultSet's meta data of a "SELECT * FROM events LIMIT 0" statement to look for the "eventParms" column. I'm also using this information_schema query in the liquibase's changelog.xml to drop the columns if all eventParms are NULL.

Comment (christianpape) [inline core/upgrade/src/main/java/org/opennms/upgrade/implementations/EventParameterMigratorOffline.java]: I found an alternative statement to check for the column's existence without referring to the schema.

**Comment by fooker** [comment]
The last two comments have been fixed. Are we good to merge?

**Comment by j-white** [comment]
I ran into a problem running the installer:
```
  Processed 5420450 eventparms entries, 30207411 event parameters inserted...                                                                                 
  Rows migrated. Dropping column 'eventparms'...                                                                                                              
  Warning: can't perform the upgrade operation because: Can't move event parameters to table: ERROR: cannot drop table events column eventparms because other 
objects depend on it                                                                                                                                          
  Detail: view all_alarms depends on table events column eventparms                                                                                           
  Hint: Use DROP ... CASCADE to drop the dependent objects too.                                                                                               
- Executing rollback phase                                                                                                                                    
- Running post-execution phase                                                                              
                                                                                                                                                              
Finished in 3 hours                                                                                                                                           
...

Upgrade completed successfully!
[jesse@noise opennms]$ sudo ./bin/install -dis


INFO 8/28/17 10:39 AM:liquibase: Reading from databasechangelog                      
                                                                                                                                
                                                                                                                                                              
SEVERE 8/28/17 10:41 AM:liquibase: Change Set 21.0.0/changelog.xml::21.0.0-event-parameters-remove-column::cpape failed.  Error: Error executing SQL ALTER TAB
LE events DROP COLUMN eventparms: ERROR: cannot drop table events column eventparms because other objects depend on it                                        
  Detail: view all_alarms depends on table events column eventparms                                                                                         
  Hint: Use DROP ... CASCADE to drop the dependent objects too.                                                                                               
liquibase.exception.DatabaseException: Error executing SQL ALTER TABLE events DROP COLUMN eventparms: ERROR: cannot drop table events column eventparms becaus
e other objects depend on it                                                                                
  Detail: view all_alarms depends on table events column eventparms                             
  Hint: Use DROP ... CASCADE to drop the dependent objects too.                                                                                               
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:62)                    
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:104)                                                     
        at liquibase.database.AbstractDatabase.execute(AbstractDatabase.java:1091)                                                
        at liquibase.database.AbstractDatabase.executeStatements(AbstractDatabase.java:1075)                                                                 
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:317)                                                                                       
        at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:27)                    
        at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:58)                                                                               
        at liquibase.Liquibase.update(Liquibase.java:114)                                                                                                  
        at org.opennms.core.schema.Migrator.migrate(Migrator.java:477)                                      
        at org.opennms.install.Installer.install(Installer.java:274)                                                                                        
        at org.opennms.install.Installer.main(Installer.java:990)                                                                                             
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)                                    
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)                                                                      
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)                                                              
        at java.lang.reflect.Method.invoke(Method.java:498)                                                                                                   
        at org.opennms.bootstrap.Bootstrap$4.run(Bootstrap.java:525)                                                                                       
        at java.lang.Thread.run(Thread.java:748)                             
Caused by: org.postgresql.util.PSQLException: ERROR: cannot drop table events column eventparms because other objects depend on it                            
  Detail: view all_alarms depends on table events column eventparms                                                                                        
  Hint: Use DROP ... CASCADE to drop the dependent objects too.                                        
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2458)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2158)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:291)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:432)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:358)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:305)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:291)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:269)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:265)
        at liquibase.executor.jvm.JdbcExecutor$1ExecuteStatementCallback.doInStatement(JdbcExecutor.java:92)
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:55)
        ... 16 more
INFO 8/28/17 10:41 AM:liquibase: Successfully released change log lock
java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:498)
        at org.opennms.bootstrap.Bootstrap$4.run(Bootstrap.java:525)
        at java.lang.Thread.run(Thread.java:748)
Caused by: org.opennms.core.schema.MigrationException: unable to migrate the database
        at org.opennms.core.schema.Migrator.migrate(Migrator.java:479)
        at org.opennms.install.Installer.install(Installer.java:274)
        at org.opennms.install.Installer.main(Installer.java:990)
        ... 6 more
Caused by: liquibase.exception.MigrationFailedException: Migration failed for change set 21.0.0/changelog.xml::21.0.0-event-parameters-remove-column::cpape:
     Reason: liquibase.exception.DatabaseException: Error executing SQL ALTER TABLE events DROP COLUMN eventparms: ERROR: cannot drop table events column even
tparms because other objects depend on it
  Detail: view all_alarms depends on table events column eventparms
  Hint: Use DROP ... CASCADE to drop the dependent objects too.:
          Caused By: Error executing SQL ALTER TABLE events DROP COLUMN eventparms: ERROR: cannot drop table events column eventparms because other objects de
pend on it
  Detail: view all_alarms depends on table events column eventparms
  Hint: Use DROP ... CASCADE to drop the dependent objects too.:                                                                                      [1/1929]
          Caused By: ERROR: cannot drop table events column eventparms because other objects depend on it
  Detail: view all_alarms depends on table events column eventparms
  Hint: Use DROP ... CASCADE to drop the dependent objects too.
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:347)
        at liquibase.changelog.visitor.UpdateVisitor.visit(UpdateVisitor.java:27)
        at liquibase.changelog.ChangeLogIterator.run(ChangeLogIterator.java:58)
        at liquibase.Liquibase.update(Liquibase.java:114)
        at org.opennms.core.schema.Migrator.migrate(Migrator.java:477)
        ... 8 more
Caused by: liquibase.exception.DatabaseException: Error executing SQL ALTER TABLE events DROP COLUMN eventparms: ERROR: cannot drop table events column eventp
arms because other objects depend on it
  Detail: view all_alarms depends on table events column eventparms
  Hint: Use DROP ... CASCADE to drop the dependent objects too.
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:62)
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:104)
        at liquibase.database.AbstractDatabase.execute(AbstractDatabase.java:1091)
        at liquibase.database.AbstractDatabase.executeStatements(AbstractDatabase.java:1075)
        at liquibase.changelog.ChangeSet.execute(ChangeSet.java:317)
        ... 12 more
Caused by: org.postgresql.util.PSQLException: ERROR: cannot drop table events column eventparms because other objects depend on it
  Detail: view all_alarms depends on table events column eventparms
  Hint: Use DROP ... CASCADE to drop the dependent objects too.
        at org.postgresql.core.v3.QueryExecutorImpl.receiveErrorResponse(QueryExecutorImpl.java:2458)
        at org.postgresql.core.v3.QueryExecutorImpl.processResults(QueryExecutorImpl.java:2158)
        at org.postgresql.core.v3.QueryExecutorImpl.execute(QueryExecutorImpl.java:291)
        at org.postgresql.jdbc.PgStatement.executeInternal(PgStatement.java:432)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:358)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:305)
        at org.postgresql.jdbc.PgStatement.executeCachedSql(PgStatement.java:291)
        at org.postgresql.jdbc.PgStatement.executeWithFlags(PgStatement.java:269)
        at org.postgresql.jdbc.PgStatement.execute(PgStatement.java:265)
        at liquibase.executor.jvm.JdbcExecutor$1ExecuteStatementCallback.doInStatement(JdbcExecutor.java:92)
        at liquibase.executor.jvm.JdbcExecutor.execute(JdbcExecutor.java:55)
        ... 16 more
[jesse@noise opennms]$
```

**Comment by christianpape** [comment]
I think we do not provide a view named 'all_alarms'? I can't find it on a fresh install and neither on my production system (20.0.2) ? I think it's an addition in the used database dump. Users can of course add any kind of constraints that will conflict with our code-driven or liquibase schema changes. I think install.sh should fail in this case instead of using DROP ... CASCADE because this will drop all related objects like triggers, views and stored procedures.

**Comment by j-white** [comment]
Yep, you're right. That's a relic of the database I used. In this case, the behavior makes sense.
